ops/cli.py

```diff
@@ -1,351 +1,319 @@
-
-
-def clone(url: str, branch: str):
-    """Clone the BRANCH of git repo at URL."""
-    info(f"=> Cloning {branch} of {url} ...")
-    shell(f"git clone {url} --branch {branch}")
-    success(f"[✔] Cloned {branch} of {url}.")
-
-
-def render_ctmpl(ctmpl_file: str, **kwargs) -> int:
-    """Render a ctmpl file."""
-    info("=> Rendering ctmpl file {ctmpl_file}")
-    envs = ""
-    if kwargs:
-        for k, v in kwargs.items():
-            envs += f"-e {k}={v}"
-    info(f"=> Feeding variables: {envs}")
-    command = " ".join(['docker run -i --rm -v "$(pwd)":/working',
-                        '-v "$HOME"/.vault-token:/root/.vault-token',
-                        f'{envs}',
-                        'broadinstitute/dsde-toolbox:dev',
-                        '/usr/local/bin/render-ctmpls.sh -k',
-                        f'"{ctmpl_file}"'])
-    shell_unchecked(command)
-    success(f"[✔] Rendered file {ctmpl_file.split('.ctmpl')[0]}")
-
-
-def commands_available():
-    commands = ["docker", "gcloud", "git", "helm", "java", "kubectl", "jq"]
-    for cmd in commands:
-        if not shutil.which(cmd):
-            error(f"=> {cmd} is not in PATH. Please install it!")
-
-
-def set_up_helm():
-    info("=> Setting up Helm charts")
-    ADD = " ".join(["helm repo add gotc-charts",
-                    "https://broadinstitute.github.io/gotc-helm-repo/"])
-    shell(ADD)
+@dataclass
+class WflInstanceConfig:
+    """Data class representing the 'state' of this script's configuration as it interacts with infrastructure."""
+    command: str
+    environment: str
+    instance_id: str
+    dry_run: bool = False
+    version: str = None
+    project: str = None
+    cloud_sql_name: str = None
+    cloud_sql_local_proxy_container: str = None
+    cluster_zone: str = None
+    cluster_name: str = None
+    cluster_namespace: str = None
+    db_username: str = None
+    db_password: str = None
+    rendered_values_file: str = None
+
+
+def infer_missing_arguments_pre_validate(config: WflInstanceConfig) -> None:
+    """Infer and store all arguments required for validation steps."""
+    if not config.project:
+        info("=>  Inferring project name to be `broad-{ENV}`")
+        config.project = f"broad-{config.environment}"
+
+
+def validate_cloud_sql_name(config: WflInstanceConfig) -> None:
+    """Validate that the stored Cloud SQL name exists."""
+    if config.cloud_sql_name:
+        info("=>  Validating Cloud SQL name")
+        instances = json.loads(shell(f"gcloud --project {config.project} --format=json "
+                                     "sql instances list"))
+        if config.cloud_sql_name not in [i["name"] for i in instances]:
+            info(f"    Available Cloud SQL instances in {config.project}:", plain=True)
+            info(shell(f"gcloud --project {config.project} sql instances list", quiet=True), plain=True)
+            raise RuntimeError(f"Cloud SQL instance {config.cloud_sql_name} not found")
```

Comment:
Just thinking what the user will see there (and in other places where yu've `info`ed before `raise`)- will we see diagnostics before the error message? That might be a bit confusing if so.
Maybe bundle the diagnostics with the error message?

Reply:
This is an issue latent in using exceptions for control flow that I pointed out earlier--that doing the messages myself would provide better output. I'd like to avoid printing `gcloud`'s entire output in red text to stderr, but it is either that or print this first if we choose to use exceptions.

Reply:
Then use an exception type that has some "details" field?

Reply:
Going to back to printing it myself.